### PR TITLE
Add function serializeSingleColumn to PrestoVectorSerde

### PR DESCRIPTION
**Pull request by pramodsatya**
https://github.com/prestodb/presto/pull/23331 adds a native expression optimizer to delegate expression evaluation to the native sidecar. This is used to constant fold expressions on the presto native sidecar, instead of on the presto java coordinator (which is the current behavior). https://github.com/prestodb/presto/pull/22927 implements a proxygen endpoint to accept `RowExpression`s from `NativeSidecarExpressionInterpreter`, optimize them if possible (rewrite special form expressions), and compile the `RowExpression` to a velox expression with constant folding enabled. This velox expression is then converted back to a `RowExpression` and returned by the sidecar to the coordinator.

When the constant folded velox expression is of type `velox::exec::ConstantExpr`, we need to return a `RowExpression` of type `ConstantExpression`. This requires us to serialize the constant value from `velox::exec::ConstantExpr` into `protocol::ConstantExpression::valueBlock`. This can be done by serializing the constant value vector to presto SerializedPage::column format, followed by base 64 encoding the result (reverse engineering the logic from `Base64Util.cpp::readBlock`). 

This PR adds a new function, `serializeSingleColumn`, to `PrestoVectorSerde`. This can be used to serialize input data from vectors containing a single element into a single PrestoPage column format (without the PrestoPage header). 
This function is not added to `PrestoBatchVectorSerializer` alongside the existing `serialize` function since that would require adding it as a virtual function in `BatchVectorSerializer` as well, and this is not desirable since the `PrestoPage` format is not relevant in this base class. There is an existing function `deserializeSingleColumn` in `PrestoVectorSerde` which is used to deserialize data from a single column, since `serializeSingleColumn` performs the inverse operation to this function, it is added alongside it in `PrestoVectorSerde`.

### CODE REVIEW

**Comment by aditi-pandit** [comment]
@pramodsatya : The changes look okay. But please can you explain what specifically in the linked PR needed these changes. I do recall our discussion. But it would be good for other reviewers as well.

**Comment by pramodsatya** [comment]
> @pramodsatya : The changes look okay. But please can you explain what specifically in the linked PR needed these changes. I do recall our discussion. But it would be good for other reviewers as well.

Thanks @aditi-pandit, updated the PR description to explain why this change is needed.

File: velox/serializers/PrestoSerializer.cpp [hunk 1]
```diff
@@ -4261,6 +4190,113 @@ void PrestoVectorSerde::deserializeSingleColumn(
   *result = row->childAt(0);
 }
 
+void PrestoVectorSerde::serializeColumn(
```
Comment (aditi-pandit) [inline velox/serializers/PrestoSerializer.cpp:4193]: @pramodsatya : Is this new code ?

Comment (pramodsatya) [reply velox/serializers/PrestoSerializer.cpp:4193]: No @aditi-pandit, this function was present [here](https://github.com/facebookincubator/velox/blob/763c19c0c767035ce03573ca2dc852b1303eeb6f/velox/serializers/PrestoSerializer.cpp#L2256) and is now added to `PrestoVectorSerde`. 
The second `serializeColumn` function was present [here](https://github.com/facebookincubator/velox/blob/763c19c0c767035ce03573ca2dc852b1303eeb6f/velox/serializers/PrestoSerializer.cpp#L2256), and is also added to `PrestoVectorSerde`.

**Comment by aditi-pandit** [comment]
@kevinwilfong : What do you think of these changes ? Please can you help with this review.

**Comment by pedroerp** [comment]
Hi @pramodsatya, not sure if I'm following. Why can't you just serialize the results of the expression (which are probably encoded as constant vectors) as PrestoPage using this API? It will maintain the encoding, then they can be deserialized by the Presto Java code. 

> https://github.com/facebookincubator/velox/blob/main/velox/serializers/PrestoSerializer.h#L41-L43

**Comment by pramodsatya** [comment]
> Hi @pramodsatya, not sure if I'm following. Why can't you just serialize the results of the expression (which are probably encoded as constant vectors) as PrestoPage using this API? It will maintain the encoding, then they can be deserialized by the Presto Java code.
> 
> > https://github.com/facebookincubator/velox/blob/main/velox/serializers/PrestoSerializer.h#L41-L43

Hi @pedroerp, thanks for the suggestion. Initially I did try to use the `serialize` function in `PrestoBatchVectorSerializer`, which serializes the velox vector in PrestoPage format, to serialize the constant encoded vector from `exec::ConstantExpr` into `protocol::Block::data` (for constructing the `protocol::ConstantExpression::valueBlock`). However, this was found to result in a deserialization error on the Presto coordinator. Upon further investigation, we noticed this was because the Presto coordinator expects just the serialized column from PrestoPage, without the header, in the field `protocol::ConstantExpression::valueBlock::data`. 
We also noticed that in `PrestoToVeloxExpr.cpp`, the `valueBlock` from Presto `protocol::ConstantExpression` is converted to velox vector by deserializing just the column from PrestoPage (without the header) in function `VeloxExprConverter::getConstantValue()` -> `deserializeSingleColumn` (we are attempting to do an inverse operation to this function when constructing a Presto ConstantExpression from velox ConstantExpr here).

Using `serializeColumn` to construct `valueBlock` in `protocol::ConstantExpression` fixed the error (ref: function `RowExpressionEvaluator::getConstantRowExpression` in [`RowExpressionEvaluator.cpp`](https://github.com/prestodb/presto/pull/22927/files#diff-834382a712c4bc5d544451fd6bca575fa1efae0e57b9cde81138ed5874797139)) and we were able to test it end to end. We were also able to solve the serialization error by just skipping over the fixed number of bytes occupied by the header in PrestoPage, but since there already is a `serializeColumn` function present in the codebase, we figured it would be better to expose it for use via velox. 

Please let me know if I am missing something.

**Comment by pedroerp** [comment]
@pramodsatya thank you for clarifying. Could we instead provide an option in PrestoOptions to make the serializer skip the header instead?

I think what I'm slightly concerned is moving VectorStream and all that stuff to the API (header) of this class, making the client more verbose, and the API a bit less intuitive.

File: velox/serializers/PrestoSerializer.cpp [hunk 2]
```diff
@@ -4265,6 +4265,26 @@ void PrestoVectorSerde::deserializeSingleColumn(
   *result = row->childAt(0);
 }
 
+void PrestoVectorSerde::serializeSingleColumn(
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/serializers/PrestoSerializer.cpp:4266]: `deserializeSingleColumn` can be used to deserialize the results of this serialization, updated the function's comment to indicate the same.

File: velox/serializers/PrestoSerializer.cpp [hunk 3]
```diff
@@ -4265,6 +4265,26 @@ void PrestoVectorSerde::deserializeSingleColumn(
...
+    const VectorPtr& vector,
+    const PrestoOptions& opts,
+    memory::MemoryPool* pool,
+    std::ostream* output) {
+  const auto numRows = vector->size();
```
[the remark this answers is not in the document]
Comment (pramodsatya) [reply velox/serializers/PrestoSerializer.cpp]: Thanks for pointing it out, renamed the variable to size and also removed the condition that the input vector contain a single element since it was specific to our usecase in Presto.

**Review by pramodsatya — COMMENTED** [review]
Thanks for the feedback @aditi-pandit, addressed the comments and modified the existing unit-test for `deserializeSingleColumn` to test `serializeSingleColumn` as well. Could you please take another look?

File: velox/serializers/tests/PrestoSerializerTest.cpp [hunk 4]
```diff
@@ -1511,34 +1511,14 @@ INSTANTIATE_TEST_SUITE_P(
...
-    validateLexer(output.str(), {});
+    serde_->serializeSingleColumn(vector, nullptr, pool_.get(), &output);
+    const auto serialized = output.str();
 
-    // Remove the PrestoPage header and Number of columns section from the
-    // serialized data.
-    std::string input = output.str().substr(kBytesToTrim);
-
-    auto byteStream = toByteStream(input);
+    auto byteStream = toByteStream(serialized);
```
Comment (aditi-pandit) [inline velox/serializers/tests/PrestoSerializerTest.cpp:1521]: Why did you remove the validateLexer call ?

File: velox/serializers/PrestoSerializer.cpp [hunk 5]
```diff
@@ -4263,6 +4263,34 @@ void PrestoVectorSerde::deserializeSingleColumn(
...
+    const Options* opts,
+    memory::MemoryPool* pool,
+    std::ostream* output) {
+  const auto prestoOptions = toPrestoOptions(opts);
+  VELOX_USER_CHECK_EQ(
+      prestoOptions.compressionKind,
+      common::CompressionKind::CompressionKind_NONE);
+  VELOX_USER_CHECK_EQ(prestoOptions.nullsFirst, false);
+
+  const IndexRange range{0, vector->size()};
```
Comment (aditi-pandit) [inline velox/serializers/PrestoSerializer.cpp:4277]: Just confirming that vector->size() can be >1 ? If there is any restriction for vector length then it should be validated.

**Review by aditi-pandit — COMMENTED** [review]
Thanks @pramodsatya. Looks good minus the comments.

File: velox/serializers/PrestoSerializer.cpp (hunk 5 above)
Comment (pramodsatya) [reply velox/serializers/PrestoSerializer.cpp:4277]: Yes, there is no restriction on vector->size() and it can be greater than 1. This check was mistakenly added before, based on the constant folding Prestissimo PR usecase, and it is not needed in velox.

File: velox/serializers/tests/PrestoSerializerTest.cpp (hunk 4 above)
Comment (pramodsatya) [reply velox/serializers/tests/PrestoSerializerTest.cpp:1521]: `validateLexer` call was removed here because the `PrestoPage` header is no longer present in the serialized data obtained from the call to `serializeSingleColumn`, and this results in the following error from `validateLexer`:
```
velox/velox/serializers/tests/PrestoSerializerTest.cpp:177: Failure
Value of: status.ok()
  Actual: false
Expected: true
PrestoPage header invalid
```
This wasn't an issue before because the function `serialize` was used to obtain the serialized data including the `PrestoPage` header, and the header bytes were just skipped over to test the function `deserializeSingleColumn`. Please let me know if this is fine.

Comment (aditi-pandit) [reply velox/serializers/tests/PrestoSerializerTest.cpp:1521]: Then it might be better to test both code paths (the original one with serialize and your new one using serializeColumn) separately, so that the distinction is clear. Lets not completely overwrite the original test.

Comment (pramodsatya) [reply velox/serializers/tests/PrestoSerializerTest.cpp:1552]: Thanks for the suggestion @aditi-pandit, added another lambda, `testSerializeSingleColumnRoundTrip`, to test `serializeSingleColumn` and `deserializeSingleColumn` round trip without modifying the existing code path. Could you please take another look?

File: velox/serializers/tests/PrestoSerializerTest.cpp [hunk 6]
```diff
@@ -1545,6 +1544,19 @@ TEST_F(PrestoSerializerTest, deserializeSingleColumn) {
     assertEqualVectors(vector, deserialized);
   };
 
+  auto testSerializeSingleColumnRoundTrip = [&](const VectorPtr& vector) {
+    std::ostringstream output;
+    serde_->serializeSingleColumn(vector, nullptr, pool_.get(), &output);
+    const auto serialized = output.str();
+
+    auto byteStream = toByteStream(serialized);
```
Comment (aditi-pandit) [inline velox/serializers/tests/PrestoSerializerTest.cpp:1552]: Abstract this common code between the 2 lamdbas as a single function and re-use to indicate that the deserializeSingleColumn is with the same function.

File: velox/serializers/tests/PrestoSerializerTest.cpp [hunk 7]
```diff
@@ -1511,14 +1511,13 @@ INSTANTIATE_TEST_SUITE_P(
         common::CompressionKind::CompressionKind_LZ4,
         common::CompressionKind::CompressionKind_GZIP));
 
-TEST_F(PrestoSerializerTest, deserializeSingleColumn) {
-  // Verify that deserializeSingleColumn API can handle all supported types.
+TEST_F(PrestoSerializerTest, serdeSingleColumn) {
   static const size_t kPrestoPageHeaderBytes = 21;
```
Comment (aditi-pandit) [inline velox/serializers/tests/PrestoSerializerTest.cpp:1515]: Move these constants in the testDeserializeSingleColumnRoundTrip lambda

File: velox/serializers/tests/PrestoSerializerTest.cpp [hunk 8]
```diff
@@ -1511,14 +1511,13 @@ INSTANTIATE_TEST_SUITE_P(
...
   static const size_t kNumOfColumnsSerializedBytes = sizeof(int32_t);
   static const size_t kBytesToTrim =
       kPrestoPageHeaderBytes + kNumOfColumnsSerializedBytes;
 
-  auto testRoundTripSingleColumn = [&](const VectorPtr& vector) {
+  auto testDeserializeSingleColumnRoundTrip = [&](const VectorPtr& vector) {
```
Comment (aditi-pandit) [inline velox/serializers/tests/PrestoSerializerTest.cpp:1520]: Maybe call the lambdas testSerializeRoundTrip and testSerializeSingleColumnRoundTrip to make the difference between the 2 lambdas more clear. Also write a comment to indicate that between serialize and serializeSIngleColumn is the PrestoPage header and number of columns section from the serialized data.

File: velox/serializers/tests/PrestoSerializerTest.cpp (hunk 6 above)
Comment (pramodsatya) [reply velox/serializers/tests/PrestoSerializerTest.cpp]: Thanks @aditi-pandit, updated accordingly. Could you please take another look?

File: velox/serializers/tests/PrestoSerializerTest.cpp [hunk 9]
```diff
@@ -765,6 +765,16 @@ class PrestoSerializerTest
          makeMapVector<StringView, int32_t>({})});
   }
 
+  void testDeserializeSingleColumn(
+      const std::string& serializedData,
+      const VectorPtr& vector) {
```
Comment (aditi-pandit) [inline velox/serializers/tests/PrestoSerializerTest.cpp:770]: Nit : Rename variable to expected.

**Review by aditi-pandit — APPROVED** [review]
Thanks @pramodsatya. Minor comment.

**Review by pedroerp — APPROVED** [review]
Thanks @pramodsatya
